From b1418d41efec8c25f33de0caf9fb7cc979463290 Mon Sep 17 00:00:00 2001
Message-Id: <b1418d41efec8c25f33de0caf9fb7cc979463290.1413837857.git.chang-joon.lee@intel.com>
In-Reply-To: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
References: <006497b7323eca32c2f489a1a9f662743c0d16d2.1413837855.git.chang-joon.lee@intel.com>
From: Akash Goel <akash.goel@intel.com>
Date: Thu, 2 Jan 2014 12:47:40 +0530
Subject: [PATCH 128/312] FOR_UPSTREAM [VPG]: drm/i915/vlv: Not reallocating
 VLV PCTX upon every suspend/resume.

VLV PCTX will come from stolen memory. Upon every suspend/resume cycle,
this is being deallocated/reallocated. Given that PCTX has to be there
at a constant stolen memory offset, doing this is not required. Also
there is a chance of it getting corrupted, if this range gets used for
some other allocation(on resume), which can result in GPU hangs.

Issue: VIZ-3428
Change-Id: Id3bf32e44c148422f42204873512bd1be173f4a9
Signed-off-by: Akash Goel <akash.goel@intel.com>
Signed-off-by: Deepak S <deepak.s@intel.com>
Reviewed-on: https://android.intel.com/164326
Tested-by: sys_abtbuild <sys_abtbuild@intel.com>
Reviewed-by: Bloomfield, Jon <jon.bloomfield@intel.com>

Conflicts:

	drivers/gpu/drm/i915/intel_pm.c

Signed-off-by: Jon Bloomfield <jon.bloomfield@intel.com>
---
 drivers/gpu/drm/i915/i915_gem_stolen.c |    7 +++++++
 drivers/gpu/drm/i915/intel_pm.c        |    4 ++++
 2 files changed, 11 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_gem_stolen.c b/drivers/gpu/drm/i915/i915_gem_stolen.c
index d217f26..1bb4332 100644
--- a/drivers/gpu/drm/i915/i915_gem_stolen.c
+++ b/drivers/gpu/drm/i915/i915_gem_stolen.c
@@ -206,6 +206,13 @@ void i915_gem_cleanup_stolen(struct drm_device *dev)
 	if (!drm_mm_initialized(&dev_priv->mm.stolen))
 		return;
 
+	if (dev_priv->vlv_pctx) {
+		mutex_lock(&dev->struct_mutex);
+		drm_gem_object_unreference(&dev_priv->vlv_pctx->base);
+		mutex_unlock(&dev->struct_mutex);
+		dev_priv->vlv_pctx = NULL;
+	}
+
 	i915_gem_stolen_cleanup_compression(dev);
 	drm_mm_takedown(&dev_priv->mm.stolen);
 }
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index bab3738..77fd033 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -4182,6 +4182,10 @@ static void valleyview_setup_pctx(struct drm_device *dev)
 	u32 pcbr;
 	int pctx_size = 24*1024;
 
+	/* If PC Context is already there, then bail out*/
+	if (dev_priv->vlv_pctx)
+		return;
+
 	WARN_ON(!mutex_is_locked(&dev->struct_mutex));
 
 	pcbr = I915_READ(VLV_PCBR);
-- 
1.7.9.5

